File: days/day_09/solutions/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;

sub solve {
    my ($input_data) = @_;
    my @lines = split /\n/, $input_data;
    @lines = grep { $_ =~ /\S/ } map { s/^\s+|\s+$//g; $_ } @lines;
    
    # Parse coordinates
    my @red_tiles = ();
    for my $line (@lines) {
        if ($line =~ /,/) {
            my @parts = split /,/, $line;
            my $x = int($parts[0]);
            my $y = int($parts[1]);
            push @red_tiles, [$x, $y];
        }
    }
    
    return ("0", "0") if @red_tiles < 2;
    
    # Part 1: Find largest rectangle area using any two red tiles as corners
    my $max_area_part1 = 0;
    for my $i (0..$#red_tiles) {
        for my $j ($i+1..$#red_tiles) {
            my ($x1, $y1) = @{$red_tiles[$i]};
            my ($x2, $y2) = @{$red_tiles[$j]};
            my $width = abs($x1 - $x2) + 1;
            my $height = abs($y1 - $y2) + 1;
            my $area = $width * $height;
            $max_area_part1 = $area if $area > $max_area_part1;
        }
    }
    
    # Part 2: Coordinate compression + flood-fill + prefix sums
    # Collect all x and y coordinates for compression
    my %all_x_set = ();
    my %all_y_set = ();
    for my $tile (@red_tiles) {
        my ($x, $y) = @$tile;
        $all_x_set{$x} = 1;
        $all_x_set{$x + 1} = 1;
        $all_y_set{$y} = 1;
        $all_y_set{$y + 1} = 1;
    }
    
    my @all_x = sort { $a <=> $b } keys %all_x_set;
    my @all_y = sort { $a <=> $b } keys %all_y_set;
    
    # Create compression maps
    my %x_to_cx = ();
    for my $i (0..$#all_x) {
        $x_to_cx{$all_x[$i]} = $i;
    }
    my %y_to_cy = ();
    for my $i (0..$#all_y) {
        $y_to_cy{$all_y[$i]} = $i;
    }
    
    my $width = scalar @all_x;
    my $height = scalar @all_y;
    
    # Build grid: 0 = outside/invalid, 1 = inside/valid
    my @grid = map { [(0) x $height] } 0..$width-1;
    
    # Mark boundary (red + green tiles)
    for my $tile (@red_tiles) {
        my ($x, $y) = @$tile;
        if (exists $x_to_cx{$x} && exists $y_to_cy{$y}) {
            $grid[$x_to_cx{$x}][$y_to_cy{$y}] = 1;
        }
    }
    
    # Connect consecutive red tiles with green tiles
    for my $i (0..$#red_tiles) {
        my ($x1, $y1) = @{$red_tiles[$i]};
        my ($x2, $y2) = @{$red_tiles[($i + 1) % @red_tiles]};
        
        if ($x1 == $x2) {
            my $start_y = $y1 < $y2 ? $y1 : $y2;
            my $end_y = $y1 > $y2 ? $y1 : $y2;
            for my $y ($start_y..$end_y) {
                if (exists $x_to_cx{$x1} && exists $y_to_cy{$y}) {
                    $grid[$x_to_cx{$x1}][$y_to_cy{$y}] = 1;
                }
            }
        } elsif ($y1 == $y2) {
            my $start_x = $x1 < $x2 ? $x1 : $x2;
            my $end_x = $x1 > $x2 ? $x1 : $x2;
            for my $x ($start_x..$end_x) {
                if (exists $x_to_cx{$x} && exists $y_to_cy{$y1}) {
                    $grid[$x_to_cx{$x}][$y_to_cy{$y1}] = 1;
                }
            }
        }
    }
    
    # Point-in-polygon check
    sub point_in_polygon {
        my ($px, $py, $red_tiles_ref) = @_;
        my $inside = 0;
        for my $i (0..$#$red_tiles_ref) {
            my ($x1, $y1) = @{$red_tiles_ref->[$i]};
            my ($x2, $y2) = @{$red_tiles_ref->[($i + 1) % @$red_tiles_ref]};
            if (($y1 > $py) != ($y2 > $py)) {
                my $intersect_x;
                if ($y2 != $y1) {
                    $intersect_x = ($py - $y1) * ($x2 - $x1) / ($y2 - $y1) + $x1;
                } else {
                    $intersect_x = $px;
                }
                if ($px < $intersect_x) {
                    $inside = !$inside;
                }
            }
        }
        return $inside;
    }
    
    # Flood fill interior
    my $found_interior = 0;
    for my $cx (0..$width-1) {
        for my $cy (0..$height-1) {
            if (!$grid[$cx][$cy]) {
                my $orig_x = $all_x[$cx];
                my $orig_y = $all_y[$cy];
                if (point_in_polygon($orig_x, $orig_y, \@red_tiles)) {
                    # Flood fill from this point
                    my @stack = ([$cx, $cy]);
                    while (@stack) {
                        my $pos = pop @stack;
                        my ($x, $y) = @$pos;
                        if ($x < 0 || $y < 0 || $x >= $width || $y >= $height || $grid[$x][$y]) {
                            next;
                        }
                        my $orig_x_val = $all_x[$x];
                        my $orig_y_val = $all_y[$y];
                        if (point_in_polygon($orig_x_val, $orig_y_val, \@red_tiles)) {
                            $grid[$x][$y] = 1;
                            push @stack, [$x - 1, $y];
                            push @stack, [$x + 1, $y];
                            push @stack, [$x, $y - 1];
                            push @stack, [$x, $y + 1];
                        }
                    }
                    $found_interior = 1;
                    last;
                }
            }
        }
        last if $found_interior;
    }
    
    # Build 2D prefix sum for O(1) rectangle queries
    my @prefix = map { [(0) x ($height + 1)] } 0..$width;
    for my $cx (0..$width-1) {
        for my $cy (0..$height-1) {
            $prefix[$cx + 1][$cy + 1] = $prefix[$cx][$cy + 1] + $prefix[$cx + 1][$cy]
                                       - $prefix[$cx][$cy] + ($grid[$cx][$cy] ? 1 : 0);
        }
    }
    
    # Helper: get sum in rectangle [cx1, cx2] x [cy1, cy2] (inclusive)
    sub rect_sum {
        my ($cx1, $cx2, $cy1, $cy2, $prefix_ref) = @_;
        return $prefix_ref->[$cx2 + 1][$cy2 + 1] - $prefix_ref->[$cx1][$cy2 + 1]
             - $prefix_ref->[$cx2 + 1][$cy1] + $prefix_ref->[$cx1][$cy1];
    }
    
    # Generate candidates sorted by area descending
    my @candidates = ();
    for my $i (0..$#red_tiles) {
        for my $j ($i+1..$#red_tiles) {
            my ($x1, $y1) = @{$red_tiles[$i]};
            my ($x2, $y2) = @{$red_tiles[$j]};
            my $min_x = $x1 < $x2 ? $x1 : $x2;
            my $max_x = $x1 > $x2 ? $x1 : $x2;
            my $min_y = $y1 < $y2 ? $y1 : $y2;
            my $max_y = $y1 > $y2 ? $y1 : $y2;
            my $area = ($max_x - $min_x + 1) * ($max_y - $min_y + 1);
            
            if (exists $x_to_cx{$min_x} && exists $x_to_cx{$max_x} &&
                exists $y_to_cy{$min_y} && exists $y_to_cy{$max_y}) {
                my $cx1 = $x_to_cx{$min_x};
                my $cx2 = $x_to_cx{$max_x};
                my $cy1 = $y_to_cy{$min_y};
                my $cy2 = $y_to_cy{$max_y};
                push @candidates, [$min_x, $max_x, $min_y, $max_y, $area, $cx1, $cx2, $cy1, $cy2];
            }
        }
    }
    
    @candidates = sort { $b->[4] <=> $a->[4] } @candidates;
    
    # Check candidates in descending area order
    my $max_area_part2 = 0;
    for my $cand (@candidates) {
        my ($min_x, $max_x, $min_y, $max_y, $area, $cx1, $cx2, $cy1, $cy2) = @$cand;
        last if $area <= $max_area_part2;
        
        # Check if rectangle is fully contained using prefix sum
        my $valid_count = rect_sum($cx1, $cx2, $cy1, $cy2, \@prefix);
        my $expected_cells = ($cx2 - $cx1 + 1) * ($cy2 - $cy1 + 1);
        
        if ($valid_count == $expected_cells) {
            # Check corners to ensure rectangle doesn't extend beyond valid regions
            my $all_valid = 1;
            for my $corner ([$min_x, $min_y], [$min_x, $max_y], [$max_x, $min_y], [$max_x, $max_y]) {
                my ($x, $y) = @$corner;
                if (exists $x_to_cx{$x} && exists $y_to_cy{$y}) {
                    my $cx = $x_to_cx{$x};
                    my $cy = $y_to_cy{$y};
                    if (!$grid[$cx][$cy]) {
                        $all_valid = 0;
                        last;
                    }
                } else {
                    if (!point_in_polygon($x, $y, \@red_tiles)) {
                        $all_valid = 0;
                        last;
                    }
                }
            }
            
            if ($all_valid) {
                $max_area_part2 = $area;
                last;
            }
        }
    }
    
    return ($max_area_part1, $max_area_part2);
}

open my $fh, '<', "../data/input.txt" or die "Cannot open file: $!";
my $data = do { local $/; <$fh> };
close $fh;

my ($part1, $part2) = solve($data);
print "Part 1: $part1\n";
print "Part 2: $part2\n";

File: days/day_09/solutions/solution.py
from typing import List, Tuple, Set

def solve(input_data: str) -> Tuple[str, str]:
    """Solve Day 9 parts 1 and 2."""
    lines = input_data.strip().split('\n')
    lines = [line.strip() for line in lines if line.strip()]
    
    # Parse coordinates
    red_tiles = []
    for line in lines:
        if ',' in line:
            parts = line.split(',')
            x = int(parts[0])
            y = int(parts[1])
            red_tiles.append((x, y))
    
    if len(red_tiles) < 2:
        return "0", "0"
    
    # Part 1: Find largest rectangle area using any two red tiles as corners
    max_area_part1 = 0
    for i in range(len(red_tiles)):
        for j in range(i + 1, len(red_tiles)):
            x1, y1 = red_tiles[i]
            x2, y2 = red_tiles[j]
            width = abs(x1 - x2) + 1
            height = abs(y1 - y2) + 1
            area = width * height
            max_area_part1 = max(max_area_part1, area)
    
    # Part 2: Coordinate compression + flood-fill + prefix sums
    # Collect all x and y coordinates for compression
    all_x_set = set()
    all_y_set = set()
    for x, y in red_tiles:
        all_x_set.add(x)
        all_x_set.add(x + 1)
        all_y_set.add(y)
        all_y_set.add(y + 1)
    
    all_x = sorted(all_x_set)
    all_y = sorted(all_y_set)
    
    # Create compression maps
    x_to_cx = {x: i for i, x in enumerate(all_x)}
    y_to_cy = {y: i for i, y in enumerate(all_y)}
    
    width = len(all_x)
    height = len(all_y)
    
    # Build grid: False = outside/invalid, True = inside/valid
    grid = [[False] * height for _ in range(width)]
    
    # Mark boundary (red + green tiles)
    for x, y in red_tiles:
        if x in x_to_cx and y in y_to_cy:
            grid[x_to_cx[x]][y_to_cy[y]] = True
    
    # Connect consecutive red tiles with green tiles
    for i in range(len(red_tiles)):
        x1, y1 = red_tiles[i]
        x2, y2 = red_tiles[(i + 1) % len(red_tiles)]
        
        if x1 == x2:
            start_y = min(y1, y2)
            end_y = max(y1, y2)
            for y in range(start_y, end_y + 1):
                if x1 in x_to_cx and y in y_to_cy:
                    grid[x_to_cx[x1]][y_to_cy[y]] = True
        elif y1 == y2:
            start_x = min(x1, x2)
            end_x = max(x1, x2)
            for x in range(start_x, end_x + 1):
                if x in x_to_cx and y1 in y_to_cy:
                    grid[x_to_cx[x]][y_to_cy[y1]] = True
    
    # Point-in-polygon check
    def point_in_polygon(px: int, py: int) -> bool:
        inside = False
        for i in range(len(red_tiles)):
            x1, y1 = red_tiles[i]
            x2, y2 = red_tiles[(i + 1) % len(red_tiles)]
            if (y1 > py) != (y2 > py):
                intersect_x = (py - y1) * (x2 - x1) / (y2 - y1) + x1 if y2 != y1 else px
                if px < intersect_x:
                    inside = not inside
        return inside
    
    # Flood fill interior
    found_interior = False
    for cx in range(width):
        for cy in range(height):
            if not grid[cx][cy]:
                orig_x = all_x[cx]
                orig_y = all_y[cy]
                if point_in_polygon(orig_x, orig_y):
                    # Flood fill from this point
                    stack = [(cx, cy)]
                    while stack:
                        x, y = stack.pop()
                        if x >= width or y >= height or grid[x][y]:
                            continue
                        orig_x = all_x[x]
                        orig_y = all_y[y]
                        if point_in_polygon(orig_x, orig_y):
                            grid[x][y] = True
                            if x > 0:
                                stack.append((x - 1, y))
                            if x + 1 < width:
                                stack.append((x + 1, y))
                            if y > 0:
                                stack.append((x, y - 1))
                            if y + 1 < height:
                                stack.append((x, y + 1))
                    found_interior = True
                    break
        if found_interior:
            break
    
    # Build 2D prefix sum for O(1) rectangle queries
    prefix = [[0] * (height + 1) for _ in range(width + 1)]
    for cx in range(width):
        for cy in range(height):
            prefix[cx + 1][cy + 1] = (prefix[cx][cy + 1] + prefix[cx + 1][cy] 
                                      - prefix[cx][cy] + (1 if grid[cx][cy] else 0))
    
    # Helper: get sum in rectangle [cx1, cx2] x [cy1, cy2] (inclusive)
    def rect_sum(cx1: int, cx2: int, cy1: int, cy2: int) -> int:
        return (prefix[cx2 + 1][cy2 + 1] - prefix[cx1][cy2 + 1] 
                - prefix[cx2 + 1][cy1] + prefix[cx1][cy1])
    
    # Generate candidates sorted by area descending
    candidates = []
    for i in range(len(red_tiles)):
        for j in range(i + 1, len(red_tiles)):
            x1, y1 = red_tiles[i]
            x2, y2 = red_tiles[j]
            min_x = min(x1, x2)
            max_x = max(x1, x2)
            min_y = min(y1, y2)
            max_y = max(y1, y2)
            area = (max_x - min_x + 1) * (max_y - min_y + 1)
            
            if (min_x in x_to_cx and max_x in x_to_cx and 
                min_y in y_to_cy and max_y in y_to_cy):
                cx1 = x_to_cx[min_x]
                cx2 = x_to_cx[max_x]
                cy1 = y_to_cy[min_y]
                cy2 = y_to_cy[max_y]
                candidates.append((min_x, max_x, min_y, max_y, area, cx1, cx2, cy1, cy2))
    
    candidates.sort(key=lambda x: x[4], reverse=True)
    
    # Check candidates in descending area order
    max_area_part2 = 0
    for min_x, max_x, min_y, max_y, area, cx1, cx2, cy1, cy2 in candidates:
        if area <= max_area_part2:
            break
        
        # Check if rectangle is fully contained using prefix sum
        valid_count = rect_sum(cx1, cx2, cy1, cy2)
        expected_cells = (cx2 - cx1 + 1) * (cy2 - cy1 + 1)
        
        if valid_count == expected_cells:
            # Check corners to ensure rectangle doesn't extend beyond valid regions
            all_valid = True
            for x, y in [(min_x, min_y), (min_x, max_y), (max_x, min_y), (max_x, max_y)]:
                if x in x_to_cx and y in y_to_cy:
                    cx, cy = x_to_cx[x], y_to_cy[y]
                    if not grid[cx][cy]:
                        all_valid = False
                        break
                else:
                    if not point_in_polygon(x, y):
                        all_valid = False
                        break
            
            if all_valid:
                max_area_part2 = area
                break
    
    return str(max_area_part1), str(max_area_part2)


if __name__ == "__main__":
    with open("../data/input.txt", "r") as f:
        data = f.read()
    part1, part2 = solve(data)
    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")
